cupy/cusolver.py

```diff
@@ -254,3 +256,213 @@ def gesvda(a, compute_uv=True):
     u = u.reshape(a_shape[:-2] + (u.shape[-2:]))
     v = v.reshape(a_shape[:-2] + (v.shape[-2:]))
     return u, s, v
+
+
+def syevj(a, UPLO='L', with_eigen_vector=True):
+    """Eigenvalue decomposition of symmetric matrix using cusolverDn<t>syevj().
+
+    Computes eigenvalues ``w`` and (optionally) eigenvectors ``v`` of a complex
+    Hermitian or a real symmetric matrix.
+
+    Args:
+        a (cupy.ndarray): A symmetric 2-D square matrix ``(M, M)`` or a batch
+            of symmetric 2-D square matrices ``(..., M, M)``.
```

Comment:
Could you add the support for `M == 0` or mention that `M >= 1` is required?  NumPy supports empty matrices while LAPACK interface does not.  Try to be consistent in `cupy/cusolver.py`.

For `M == 0`, the line `a = a.reshape(-1, m, lda)` has to be amended, too.

Reply:
That's a good point. It looks like `numpy.linalg.eigh` indirectly requires `M >= 1` by requiring that the last two dimensions of `a` to be equal. I don't know if it's possible to create a square two dimensional array with shape `(M, M)` where `M == 0` (the closest I can think of is `numpy.array([[]])` which has shape `(1, 0)`).

```python
In [1]: import numpy as np

In [2]: np.linalg.eigh([])

    205         if a.ndim < 2:
    206             raise LinAlgError('%d-dimensional array given. Array must be '
--> 207                     'at least two-dimensional' % a.ndim)
    208

LinAlgError: 1-dimensional array given. Array must be at least two-dimensional


In [3]: np.linalg.eigh([[]])

    211         m, n = a.shape[-2:]
    212         if m != n:
--> 213             raise LinAlgError('Last 2 dimensions of the array must be square')
    214

LinAlgError: Last 2 dimensions of the array must be square
```

I can add similar checks to `cupy.linalg.eigh` and `cupy.linalg.eigvalsh`.

Does that sound good?

Reply:
We can make such an ndarray by:
```python
In [4]: np.linalg.eigh(np.zeros([0, 0]))                                                                                                                                                                           
Out[4]: (array([], dtype=float64), array([], shape=(0, 0), dtype=float64))                                                    
```

FYI.

Reply:
Ah, thanks! I didn't know that :)

Reply:
Okay, I pushed a fix for the offending line that was mentioned. Unfortunately, I will not have access to a GPU system to test locally until next Wednesday.